microservices-connector/internal/controller/gmconnector_controller.go

```diff
@@ -380,18 +426,155 @@ func (r *GMConnectorReconciler) Reconcile(ctx context.Context, req ctrl.Request)
 	//so we need to apply the router config every time
-	err := reconcileRouterService(ctx, r.Client, graph)
+	err := r.reconcileRouterService(ctx, r.Client, graph)
 	if err != nil {
 		return reconcile.Result{Requeue: true}, errors.Wrapf(err, "Failed to reconcile router service")
 	}
 
-	graph.Status.Status = fmt.Sprintf("%d/%d/%d", successService, externalService, totalService)
-	if err = r.Status().Update(context.TODO(), graph); err != nil {
-		return reconcile.Result{Requeue: true}, errors.Wrapf(err, "Failed to Update CR status to %s", graph.Status.Status)
+	err = r.collectResourceStatus(graph, ctx, int(externalService))
+	if err != nil {
+		return reconcile.Result{Requeue: true}, errors.Wrapf(err, "Failed to collect service status")
+	}
+
+	if updateExistGraph {
+		//check if the old annotations are still in the new graph
+		for k := range oldAnnotations {
+			if _, ok := graph.Status.Annotations[k]; !ok {
+				//if not, remove the resource from k8s
+				r.deleteRecordedResource(k, ctx)
+			}
+		}
+	}
+	return ctrl.Result{}, nil
+}
+
+func (r *GMConnectorReconciler) handleStatusUpdate(ctx context.Context, deployment *appsv1.Deployment) (ctrl.Result, error) {
+	for _, owner := range deployment.OwnerReferences {
+		if owner.Kind == "GMConnector" {
+			// Get the GMConnector object
+			graph := &mcv1alpha3.GMConnector{}
+			err := r.Get(ctx, types.NamespacedName{Namespace: deployment.Namespace, Name: owner.Name}, graph)
+			if err == nil {
+				externalResourceCntStr := strings.Split(graph.Status.Status, "/")[1]
+				externalResourceCnt, err := strconv.Atoi(externalResourceCntStr)
+				if err != nil {
+					fmt.Println("Error converting externalResourceCnt to int:", err)
+					return reconcile.Result{}, err
+				}
+				ue := r.collectResourceStatus(graph, ctx, externalResourceCnt)
+				if ue != nil {
+					fmt.Printf("failed to get graph before update status %s %v\n", graph.Name, err)
+					return reconcile.Result{}, err
+				}
+			}
+		}
 	}
 	return ctrl.Result{}, nil
 }
 
+func (r *GMConnectorReconciler) deleteRecordedResource(key string, ctx context.Context) {
+	kind := strings.Split(key, ":")[0]
+	apiVersion := strings.Split(key, ":")[1]
+	name := strings.Split(key, ":")[2]
+	ns := strings.Split(key, ":")[3]
+	obj := &unstructured.Unstructured{}
+	obj.SetKind(kind)
+	obj.SetName(name)
+	obj.SetNamespace(ns)
+	obj.SetAPIVersion(apiVersion)
+	err := r.Delete(ctx, obj)
+	// the resource may have been deleted by other means, i.e. user manually delete or delete namespace
+	// ignore the error if delete failed i.e resource not found
+	// since I don't want to block the process for not clearing the finalizer
+	if err != nil {
+		fmt.Printf("Failed to delete resource %s:%s:%s: %v\n", ns, kind, name, err)
+	} else {
+		fmt.Printf("Success to delete %s:%s:%s\n", ns, kind, name)
+	}
+}
+
+func (r *GMConnectorReconciler) collectResourceStatus(graph *mcv1alpha3.GMConnector, ctx context.Context, externalServiceCnt int) error {
```

Comment:
better to change this to `collectDeploymentStatus` since this collects for deployment only